### PR TITLE
Full Steering Controller

### PR DESCRIPTION
**Pull request by VivaanBahl**

### CODE REVIEW

File: LPC Projects/PWM_Sample/src/PWM_Sample.c [hunk 1]
```diff
@@ -11,42 +11,225 @@
...
+/*	Function is executed by the Callback handler after
+    a CAN message has been received */
+void CAN_rx(uint8_t msg_obj_num) {
+	/* Determine which CAN message has been received */
+	msg_obj.msgobj = msg_obj_num;
+	/* Now load up the msg_obj structure with the CAN message */
+	LPC_CCAN_API->can_receive(&msg_obj);
+	int new_desired_angle = 0;
+
+//	if (msg_obj_num == 1)
```
Comment (VivaanBahl) [inline LPC Projects/PWM_Sample/src/PWM_Sample.c]: commented code, delete or refactor

File: LPC Projects/PWM_Sample/src/PWM_Sample.c [hunk 2]
```diff
@@ -11,42 +11,225 @@
...
+	else if(msg_obj.mode_id == 0x100)
+	{
+		new_desired_angle |= (msg_obj.data[0]);
+		new_desired_angle |= (msg_obj.data[1]) << 8;
+		new_desired_angle |= (msg_obj.data[2]) << 16;
+		new_desired_angle |= (msg_obj.data[3]) << 24;
+	}
+
+	desired_steering_angle = new_desired_angle;
+//	char buf[16];
```
Comment (VivaanBahl) [inline LPC Projects/PWM_Sample/src/PWM_Sample.c]: delete or refactor

File: LPC Projects/PWM_Sample/src/PWM_Sample.c [hunk 3]
```diff
@@ -11,42 +11,225 @@
...
+    case 8:
+    case 14:
+        ticks--;
+        break;
+	}
+
+	last_state = new_state;
+	return;
+}
+///**
```
Comment (VivaanBahl) [inline LPC Projects/PWM_Sample/src/PWM_Sample.c]: double comments all the way across the sky

File: LPC Projects/PWM_Sample/src/PWM_Sample.c [hunk 4]
```diff
@@ -11,42 +11,225 @@
...
+			NULL,
+			NULL,
+		};
 	SystemCoreClockUpdate();
 	Board_Init();
+	baudrateCalculate(TEST_CCAN_BAUD_RATE, CanApiClkInitTable);
 
 	/* Enable and setup SysTick Timer at a periodic rate */
-	SysTick_Config(SystemCoreClock);
+//	SysTick_Config(SystemCoreClock / 10);
```
Comment (VivaanBahl) [inline LPC Projects/PWM_Sample/src/PWM_Sample.c]: delete or refactor

File: LPC Projects/PWM_Sample/src/PWM_Sample.c [hunk 5]
```diff
@@ -62,18 +245,83 @@ int main(void)
...
+	Chip_UART_SetupFIFOS(LPC_USART, (UART_FCR_FIFO_EN | UART_FCR_TRG_LEV2));
+	Chip_UART_TXEnable(LPC_USART);
+
+	// CAN
+	LPC_CCAN_API->init_can(&CanApiClkInitTable[0], TRUE);
+	/* Configure the CAN callback functions */
+	LPC_CCAN_API->config_calb(&callbacks);
+	/* Enable the CAN Interrupt */
+	NVIC_EnableIRQ(CAN_IRQn);
+//	NVIC_ClearPendingIRQ(CAN_IRQn);
```
Comment (VivaanBahl) [inline LPC Projects/PWM_Sample/src/PWM_Sample.c]: delete or refactor

File: LPC Projects/PWM_Sample/src/steering_driver.c [hunk 6]
```diff
@@ -115,13 +114,18 @@ void steering_set(int angle)
...
                              0,
                              MOTOR_ENCODER_TICKS_PER_REV,
                              0,
                              DEGREE_HUNDREDTHS_PER_REV);
 
-    long error = angle - actual;
+    int error = angle - actual;
+
+//	char buf[16];
+//	sprintf(buf, "%d %d %d\n", angle, actual, error);
```
Comment (VivaanBahl) [inline LPC Projects/PWM_Sample/src/steering_driver.c]: delete or refactor

File: LPC Projects/PWM_Sample/src/PWM_Sample.c [hunk 7]
```diff
@@ -11,42 +11,225 @@
...
-// * @return	Nothing
-// */
-void SysTick_Handler(void)
+#include <cr_section_macros.h>
+
+volatile uint8_t last_state;
+volatile long ticks;
+
+volatile int desired_steering_angle = 0;
+//volatile int direction = 1; // +1 forwards, -1 backwards
```
Comment (ramensandwich) [inline LPC Projects/PWM_Sample/src/PWM_Sample.c]: What was this for?  If we're not going to use it, please remove it

File: LPC Projects/PWM_Sample/src/PWM_Sample.c (hunk 1 above)
Comment (ramensandwich) [inline LPC Projects/PWM_Sample/src/PWM_Sample.c]: Is this block of dead code from when we were determining how to have multiple CAN message types?  Remove if this was solved by following block of not-dead code

File: LPC Projects/PWM_Sample/src/PWM_Sample.c [hunk 8]
```diff
@@ -62,18 +245,83 @@ int main(void)
...
+	Chip_TIMER_Reset(LPC_TIMER16_1);
+	Chip_TIMER_MatchEnableInt(LPC_TIMER16_1, 0);
+	Chip_TIMER_SetMatch(LPC_TIMER16_1, 0, 10000);
+	Chip_TIMER_ResetOnMatchEnable(LPC_TIMER16_1, 0);
+	LPC_TIMER16_1->PR = 48;
+	Chip_TIMER_Enable(LPC_TIMER16_1);
+	NVIC_ClearPendingIRQ(TIMER_16_1_IRQn);
+	NVIC_EnableIRQ(TIMER_16_1_IRQn);
+
+    // Uart crap
```
Comment (ramensandwich) [inline LPC Projects/PWM_Sample/src/PWM_Sample.c]: Please reword this

File: LPC Projects/PWM_Sample/src/steering_driver.c [hunk 9]
```diff
@@ -14,38 +14,37 @@
...
+//#define STEERING_MAX_SPEED 32000L
+#define STEERING_MAX_SPEED 1000L
+#define STEERING_KV_NUMERATOR 1L
+#define STEERING_KV_DENOMENATOR 10L
 #define STEERING_PWM_CENTER_US 1500L //The PWM value that gives no movement.
 #define STEERING_MAX_PWM 150 // max PWM value to add on top of the center_us
 #define MOTOR_ENCODER_TICKS_PER_REV 36864 // 4096 * 9 = 12-bit * 1:9 gearbox
 #define MOTOR_FEED_FORWARD 200
-#define STEERING_ERROR_THRESHOLD 5
+#define STEERING_ERROR_THRESHOLD 50
```
Comment (ramensandwich) [inline LPC Projects/PWM_Sample/src/steering_driver.c]: Why did we increase our error threshold?

File: LPC Projects/PWM_Sample/src/PWM_Sample.c (hunk 7 above)
Comment (VivaanBahl) [reply LPC Projects/PWM_Sample/src/PWM_Sample.c]: We had a sweep node in here at one point, forgot to take it out

File: LPC Projects/PWM_Sample/src/steering_driver.c (hunk 9 above)
Comment (VivaanBahl) [reply LPC Projects/PWM_Sample/src/steering_driver.c]: That was for steering pid testing, forgot to revert

File: LPC Projects/PWM_Sample/src/PWM_Sample.c [hunk 10]
```diff
@@ -11,42 +11,207 @@
...
+#define ENCODER_NVIC_NAME		EINT0_IRQn /* GPIO interrupt NVIC interrupt name */
+
+#include <cr_section_macros.h>
+
+volatile uint8_t last_state;
+volatile long ticks;
+
+volatile int desired_steering_angle = 0;
+
+int danger = 0;
```
Comment (ramensandwich) [inline LPC Projects/PWM_Sample/src/PWM_Sample.c]: This variable isn't used. Please remove

File: LPC Projects/PWM_Sample/src/PWM_Sample.c [hunk 11]
```diff
@@ -11,42 +11,207 @@
...
+		uint32_t ultrasonic_distance_mm;
+		ultrasonic_distance_mm |= (msg_obj.data[0]);
+		ultrasonic_distance_mm |= (msg_obj.data[1]) << 8;
+		ultrasonic_distance_mm |= (msg_obj.data[2]) << 16;
+		ultrasonic_distance_mm |= (msg_obj.data[3]) << 24;
+
+		if (ultrasonic_distance_mm < 1200)
+		{
+			Board_UARTPutSTR("ooooooops\n");
+			danger = 1;
```
Comment (ramensandwich) [inline LPC Projects/PWM_Sample/src/PWM_Sample.c]: danger variable not used. Please remove

File: LPC Projects/PWM_Sample/src/PWM_Sample.c [hunk 12]
```diff
@@ -62,18 +227,83 @@ int main(void)
...
+
+	msg_obj.msgobj = 0;
+	msg_obj.mode_id = 0x000;
+	msg_obj.mask = 0x000;
+	LPC_CCAN_API->config_rxmsgobj(&msg_obj);
+
+
+    /* Configure GPIO pins as input pins */
+    Chip_GPIO_SetPinDIRInput(LPC_GPIO, PIN_A_PORT, PIN_A);
+    Chip_GPIO_SetPinDIRInput(LPC_GPIO, PIN_B_PORT, PIN_B);
```
Comment (ramensandwich) [inline LPC Projects/PWM_Sample/src/PWM_Sample.c]: Being nitpicky here, but the spacing on these three lines is different from the rest of the file

**Pull request merged**